common/flatpak-dir.c

```diff
@@ -1839,6 +1841,154 @@ get_system_locations (GCancellable *cancellable,
 
+char **
+flatpak_get_preinstall_config_file_paths (GCancellable *cancellable,
+                                          GError      **error)
+{
+  g_autoptr(GPtrArray) paths = NULL;
+  g_autoptr(GFile) conf_dir = NULL;
+  g_autoptr(GFileEnumerator) dir_enum = NULL;
+  g_autoptr(GError) my_error = NULL;
+  g_autofree char *config_dir = NULL;
+
+  paths = g_ptr_array_new_with_free_func (g_free);
+  config_dir = g_strdup_printf ("%s/%s",
+                                get_config_dir_location (),
+                                SYSCONF_PREINSTALL_DIR);
+
+  if (!g_file_test (config_dir, G_FILE_TEST_IS_DIR))
+    {
+      g_info ("Skipping missing preinstall config directory %s", config_dir);
+      goto out;
+    }
+
+  conf_dir = g_file_new_for_path (config_dir);
+  dir_enum = g_file_enumerate_children (conf_dir,
+                                        G_FILE_ATTRIBUTE_STANDARD_NAME "," G_FILE_ATTRIBUTE_STANDARD_TYPE,
+                                        G_FILE_QUERY_INFO_NONE,
+                                        cancellable, &my_error);
+  if (my_error != NULL)
+    {
+      g_info ("Unexpected error retrieving preinstalls from %s: %s",
+              config_dir, my_error->message);
+      g_propagate_error (error, g_steal_pointer (&my_error));
+      return NULL;
+    }
+
+  while (TRUE)
+    {
+      GFileInfo *file_info;
+      GFile *path;
+      const char *name;
+      guint32 type;
+
+      if (!g_file_enumerator_iterate (dir_enum, &file_info, &path,
+                                      cancellable, &my_error))
+        {
+          g_info ("Unexpected error reading file in %s: %s",
+                  config_dir, my_error->message);
+          g_propagate_error (error, g_steal_pointer (&my_error));
+          return NULL;
+        }
+
+      if (file_info == NULL)
+        break;
+
+      name = g_file_info_get_attribute_byte_string (file_info, "standard::name");
+      type = g_file_info_get_attribute_uint32 (file_info, "standard::type");
+
+      if (type == G_FILE_TYPE_REGULAR && g_str_has_suffix (name, SYSCONF_PREINSTALL_FILE_EXT))
+        {
+          g_autofree char *path_str = g_file_get_path (path);
+          g_ptr_array_add (paths, g_steal_pointer (&path_str));
+        }
+    }
+
+out:
+  g_ptr_array_add (paths, NULL);
+  return (char **) g_ptr_array_free (g_steal_pointer (&paths), FALSE);
+}
+
+static gboolean
+parse_preinstall_keyfile (GKeyFile *keyfile,
+                          char    **name_out,
+                          char    **branch_out,
+                          gboolean *is_runtime_out,
+                          char    **collection_id_out,
+                          GError  **error)
+{
+  g_autofree char *name = NULL;
+  g_autofree char *branch = NULL;
+  gboolean is_runtime = FALSE;
+  g_autofree char *collection_id = NULL;
+
+  *name_out = NULL;
+  *branch_out = NULL;
+  *is_runtime_out = FALSE;
+  *collection_id_out = NULL;
+
+  if (!g_key_file_has_group (keyfile, FLATPAK_PREINSTALL_GROUP))
```

Comment:
Should an empty file really be an error?

Reply:
Oh, let me change this - thanks!